blob/azureblob/azureblob.go

```diff
@@ -0,0 +1,570 @@
+					return false
+				}
+				*p = blobInfo
+				return true
+			},
+		})
+	}
+
+	if listBlob.NextMarker.NotDone() {
+		token := uuid.New().String()
+		b.pageMarkers[token] = listBlob.NextMarker
+		page.NextPageToken = []byte(token)
+	}
+
+	if len(listBlob.Segment.BlobPrefixes) > 0 && len(listBlob.Segment.BlobItems) > 0 {
+		sort.Slice(page.Objects, func(i, j int) bool {
+			return page.Objects[i].Key < page.Objects[j].Key
+		})
+	}
+
+	return page, nil
+}
+
+func (b *bucket) SignedURL(ctx context.Context, key string, opts *driver.SignedURLOptions) (string, error) {
+
+	key = strings.Replace(key, OSPathSeparator, b.defaultDelimiter, -1)
+
+	blockBlobURL := b.urls.containerURL.NewBlobURL(key)
+	srcBlobParts := azblob.NewBlobURLParts(blockBlobURL.URL())
+	var err error
+	srcBlobParts.SAS, err = azblob.BlobSASSignatureValues{
+		Protocol:      azblob.SASProtocolHTTPS,
+		ExpiryTime:    time.Now().UTC().Add(opts.Expiry),
+		ContainerName: b.name,
+		BlobName:      key,
+		Permissions:   azblob.BlobSASPermissions{Add: true, Create: true, Delete: true, Read: true, Write: true}.String(),
+	}.NewSASQueryParameters(b.credential)
+
+	if err != nil {
+		return "", err
+	}
+
+	srcBlobURLWithSAS := srcBlobParts.URL()
+	return srcBlobURLWithSAS.String(), nil
+}
+
+type writer struct {
+	ctx         context.Context
+	urls        *serviceUrls
+	key         string
+	contentType string
+	opts        *driver.WriterOptions
+
+	w     *io.PipeWriter
+	donec chan struct{}
+	err   error
+}
+
+// NewTypedWriter implements driver.NewTypedWriter.
+func (b *bucket) NewTypedWriter(ctx context.Context, key string, contentType string, opts *driver.WriterOptions) (driver.Writer, error) {
+	containerURL := b.urls.serviceURL.NewContainerURL(b.name)
+	key = strings.Replace(key, OSPathSeparator, b.defaultDelimiter, -1)
+	blockBlobURL := containerURL.NewBlockBlobURL(key)
+
+	if opts.Metadata == nil {
+		opts.Metadata = map[string]string{}
+	}
+	if opts.BufferSize == 0 {
+		opts.BufferSize = defaultUploadBlockSize
+	}
+
+	w := &writer{
+		ctx:         ctx,
+		key:         key,
+		contentType: contentType,
+		opts:        opts,
+		urls: &serviceUrls{
+			serviceURL:   b.urls.serviceURL,
+			containerURL: &containerURL,
+			blockBlobURL: &blockBlobURL,
+		},
+		donec: make(chan struct{}),
+	}
+
+	return w, nil
+}
+
+// Write appends p to w. User must call Close to close the w after done writing.
+func (w *writer) Write(p []byte) (int, error) {
+	if w.w == nil {
+		if err := w.open(); err != nil {
+			return 0, err
+		}
+	}
+
+	select {
+	case <-w.donec:
+		return 0, w.err
+	default:
+	}
+
+	return w.w.Write(p)
+}
+
+func (w *writer) open() error {
+	pr, pw := io.Pipe()
+	w.w = pw
+
+	go func() {
+		defer close(w.donec)
+
+		var blobHTTPHeaders = azblob.BlobHTTPHeaders{}
+		if w.contentType != "" {
+			blobHTTPHeaders.ContentType = w.contentType
+		}
+		if len(w.opts.ContentMD5) > 0 {
+			blobHTTPHeaders.ContentMD5 = w.opts.ContentMD5
+		}
+
+		buf, err := ioutil.ReadAll(pr)
```

Comment:
You shouldn't have to do this (read into a buffer). Just pass `pr` directly to `azblob.UploadStreamToBlockBlob`. If you follow my suggestion above, `pr` may be `nil`, in which cases you should pass `http.NoBody`.

This allows `UploadStreamToBlockBlob` to do chunking without requiring the whole blob to be in memory. For example, an application may be streaming a large file from disk and doing a bunch of `Write` calls for it; if `pr` is given directly to `azblob`, it can send the data as it receives it (in chunks) rather than you buffering it entirely in `buf`.

Reply:
This is fixed, see changeset  [5b7786d](https://github.com/google/go-cloud/pull/936/commits/5b7786d9b71625fc3cf5ab73beceb034c059dd49)